### PR TITLE
Migration redesign

### PR DESCRIPTION
**Pull request by fmorency**
Migration integration

All unit/integration/e2e/resiliency tests are passing.

1. Migrations are specified using a configuration file
2. Every migration supported by the binary needs to be **explicitly** defined in the configuration file
3. The legacy code (pre-migration) will run if no migration configuration file is provided, even on new storage
4. The account count data attribute migration is tested
5. A dummy hotfix migration is tested
6. Migration configuration file checks are tested
7. Migrations can be disabled using the configuration file
8. The user is responsible for everything migration-related

2, 3, and 8 should be improved in future PRs.

### CODE REVIEW

File: src/many-ledger/src/migration.rs [hunk 1]
```diff
@@ -1,76 +1,16 @@
...
-impl Eq for Box<dyn Migration> {}
+pub mod block_9400;
+pub mod data;
 
-impl PartialOrd<Box<dyn Migration>> for Box<dyn Migration> {
-    fn partial_cmp(&self, other: &Box<dyn Migration>) -> Option<std::cmp::Ordering> {
-        self.name().partial_cmp(other.name())
-    }
-}
+pub type LedgerMigrations<'a> = BTreeMap<&'a str, Migration<'a, merk::Merk, ManyError>>;
```
Comment (hansl) [inline src/many-ledger/src/migration.rs:9]: Just make this `'static`. Will save you a lot of references around (and reduce your delta quite a bit). WDYT?

**Comment by hansl** [comment]
I don't understand the UX for IntelliJ PR :/

**Review by hansl — CHANGES_REQUESTED** [review]
The `'static` change would go a long way in making this easier. Just a first round to get it started.

**Review by hansl — CHANGES_REQUESTED** [review]

**Review by hansl — APPROVED** [review]

File: src/many-ledger/src/main.rs [hunk 2]
```diff
@@ -182,19 +206,48 @@ fn main() {
...
+    info!("Loading migrations from {migrations_config:?}");
+    let maybe_migrations = migrations_config.map(|file| {
+        let content = std::fs::read_to_string(file)
+            .expect("Could not read file passed to --migrations_config");
+        let mut migrations =
+            load_migrations(&MIGRATIONS, &content).expect("Could not read migration file");
+
+        // Check if we have the right number of migrations
+        if migrations.len() < MIGRATIONS.len() {
+            panic!("Migration configuration file is missing migration(s)");
```
[the remark this answers is not in the document]
Comment (fmorency) [reply src/many-ledger/src/main.rs:219]: Let's do that in another PR. I'll log an issue.

**Pull request merged**